### PR TITLE
Add support for OIDC connection strategy

### LINKED ISSUE #212: Unsupported connection strategy oidc
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The OpenID Connect provider is not supported. When it is used it throws the following error

`400 Bad Request: options.client_id is required`

### New or Affected Resource(s)

* auth0_connection

### Potential Terraform Configuration

```hcl
resource "auth0_connection" "my_con" {
  name                 = "my_con"
  strategy             = "oidc"
  is_domain_connection = false
  enabled_clients      = ["xxx"]

  options {
    client_id     = "xxx"
    client_secret = "xxx"
    scopes        = ["openid", "profile", "email"]

    configuration = {
      type                   = "back_channel"
      issuer                 = "https://xxx.com",
      jwks_uri               = "https://xxx.com/.well-known/jwks.json",
      discovery_url          = "https://xxx.com/.well-known/openid-configuration",
      token_endpoint         = "https://xxx.com/oauth2/token",
      userinfo_endpoint      = "https://xxx.com/oauth2/userInfo",
      authorization_endpoint = "https://xxx.com/oauth2/authorize"
    }
  }
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by incorrupt**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #212

Changes proposed in this pull request:

* Support the OIDC connection strategy

Output from acceptance testing:

```
$ make testacc TESTS=TestAccConnectionOIDC
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (6.44s)
PASS
coverage: 10.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	7.602s	coverage: 10.0% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.764s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.544s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]
```

### CODE REVIEW

**Comment by incorrupt** [comment]
@alexkappa any chance this can be merged soon?

Thank you!

File: auth0/structure_auth0_connection.go [hunk 1]
```diff
@@ -106,6 +106,27 @@ func flattenConnectionOptionsSMS(o *management.ConnectionOptionsSMS) interface{}
...
+		"type":                   o.GetType(),
+		"scope":                  o.GetScope(),
+		"issuer":                 o.GetIssuer(),
+		"jwks_uri":               o.GetJWKSURI(),
+		"discovery_url":          o.GetDiscoveryURL(),
+		"token_endpoint":         o.GetTokenEndpoint(),
+		"userinfo_endpoint":      o.GetUserInfoEndpoint(),
+		"authorization_endpoint": o.GetAuthorizationEndpoint(),
+
+		//"configuration": Map(d, "configuration"), // does not get read back
```
Comment (alexkappa) [inline auth0/structure_auth0_connection.go:126]: Can we remove this? It's only used for `auth0` connections iirc.

File: auth0/resource_auth0_connection.go [hunk 2]
```diff
@@ -367,6 +367,48 @@ func newConnection() *schema.Resource {
...
 							Elem:     &schema.Schema{Type: schema.TypeString},
 						},
+
+						// OIDC options
+						"type": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "",
+						},
+						"scope": {
```
Comment (alexkappa) [inline auth0/resource_auth0_connection.go:377]: Can we use `scopes` for this? With the SDK all the oauth2/openid connections have a SetScopes method to deal with this easily. You could have a look at the google connection how that does it.

File: website/docs/r/connection.html.md [hunk 3]
```diff
@@ -134,6 +134,19 @@ Arguments accepted by this resource include:
...
+* `client_secret` - (Optional) String, Case-sensitive. Client secret given by your OIDC provider.
+* `type` - (Optional) String. Value must be `back_channel` or `front_channel`
+* `scope` - (Optional) String. Value must be a space separated list of scopes. For example `openid profile email`
+* `issuer` - (Optional) String. URL of the issuer.
+* `discovery_url` - (Optional) String. Usually an URL ending with `/.well-known/openid-configuration`
+* `jwks_uri` - (Optional) String.
+* `token_endpoint` - (Optional) String.
+* `userinfo_endpoint` - (Optional) String.
+* `authorization_endpoint` - (Optional) String.
+
```
Comment (alexkappa) [inline website/docs/r/connection.html.md:149]: This is awesome, thanks for going the extra mile!

File: auth0/resource_auth0_connection_test.go [hunk 4]
```diff
@@ -255,6 +255,64 @@ resource "auth0_connection" "azure_ad" {
...
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.3154807651", "api.example.com"),
+
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scope", "openid profile email"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.type", "back_channel"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.issuer", "https://api.login.yahoo.com"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.jwks_uri", "https://api.login.yahoo.com/openid/v1/certs"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.discovery_url", "https://api.login.yahoo.com/.well-known/openid-configuration"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.token_endpoint", "https://api.login.yahoo.com/oauth2/get_token"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.userinfo_endpoint", "https://api.login.yahoo.com/openid/v1/userinfo"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.authorization_endpoint", "https://api.login.yahoo.com/oauth2/request_auth"),
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:285]: We should also have some update tests. It's very often we catch a number of issues when having forgotten the update operation.

File: auth0/resource_auth0_connection.go (hunk 2 above)
Comment (alexkappa) [reply auth0/resource_auth0_connection.go:377]: Also see the `expandConnectionOptionsScopes` I believe its relevant to this.

Comment (incorrupt) [reply auth0/resource_auth0_connection.go:389]: Thanks for your review. First things first, this is the first time I touch Go so while I am able to reproduce stuff that already exists, it's quite challenging for me to come up with new stuff!

I saw that setScopes function but [it's not present for the OIDC connection in the SDK](https://github.com/go-auth0/auth0/blob/9fd4a3c6d5ac94db0161db19ee5d16b0e04216ca/management/connection.go#L481), it seems to be the only one connection in this case.

Comment (alexkappa) [reply auth0/resource_auth0_connection.go:389]: No worries at all! I'm happy to help. If you are having any trouble just let me know.

You are right, I must have missed a bunch `SetScopes` in go-auth0/auth0. I can do that from tomorrow/weekend or if you're in a hurry, feel free to patch it yourself. You can use the same one as Salesforce above OIDC.

You can use the `GoogleOAuth2` as your guiding principle. Also #221 and #222 are great examples, very similar to your PR.

Comment (incorrupt) [reply auth0/resource_auth0_connection.go:389]: I was about to do it but am not sure how to test it and have the `management.gen.go` generated (I suppose it is)... So I prefer to leave it to you.

**Comment by incorrupt** [comment]
@alexkappa I'm stuck on the tests. I get the following:
`TestAccConnectionOIDC: testing.go:674: Step 0 error: Check failed: Check 16/17 error: auth0_connection.oidc: Attribute 'options.0.scopes.881205744' not found`
How do I find out the right attribute number?

**Comment by alexkappa** [comment]
Hi Romain, to discover what key to use you can use the debug.DumpAttr()
function from the debug package. Use it as one of your checks and it will
print the entire state.

To view the output, you’ll need to set the TF_LOG=DEBUG env var when
running tests.

File: auth0/resource_auth0_connection_test.go [hunk 5]
```diff
@@ -255,6 +255,110 @@ resource "auth0_connection" "azure_ad" {
...
+			"example.com"
+		]
+		type                   = "front_channel"
+		issuer                 = "https://www.paypalobjects.com"
+		jwks_uri               = "https://api.paypal.com/v1/oauth2/certs"
+		discovery_url          = "https://www.paypalobjects.com/.well-known/openid-configuration"
+		token_endpoint         = "https://api.paypal.com/v1/oauth2/token"
+		userinfo_endpoint      = "https://api.paypal.com/v1/oauth2/token/userinfo"
+		authorization_endpoint = "https://www.paypal.com/signin/authorize"
+		scopes = [ "email", "profile" ]
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:357]:
```suggestion
		scopes = [ "email", "openid" ]
```

File: auth0/resource_auth0_connection_test.go [hunk 6]
```diff
@@ -255,6 +255,110 @@ resource "auth0_connection" "azure_ad" {
...
+				Config: random.Template(testAccConnectionOIDCConfig, rand),
+				Check: resource.ComposeTestCheckFunc(
+					random.TestCheckResourceAttr("auth0_connection.oidc", "name", "Acceptance-Test-OIDC-{{.random}}", rand),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "strategy", "oidc"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.client_id", "123456"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.client_secret", "123456"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.#", "2"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.3506632655", "example.com"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.3154807651", "api.example.com"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scope", "openid profile email"),
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:277]:
```suggestion
```

File: auth0/resource_auth0_connection_test.go [hunk 7]
```diff
@@ -255,6 +255,110 @@ resource "auth0_connection" "azure_ad" {
...
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.jwks_uri", "https://api.login.yahoo.com/openid/v1/certs"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.discovery_url", "https://api.login.yahoo.com/.well-known/openid-configuration"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.token_endpoint", "https://api.login.yahoo.com/oauth2/get_token"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.userinfo_endpoint", "https://api.login.yahoo.com/openid/v1/userinfo"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.authorization_endpoint", "https://api.login.yahoo.com/oauth2/request_auth"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.#", "3"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.881205744", "email"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.4080487570", "profile"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.862208977", "openid"),
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:288]:
```suggestion
					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.2517049750", "openid"),
					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.4080487570", "profile"),
					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.881205744", "email"),
```

File: auth0/resource_auth0_connection_test.go [hunk 8]
```diff
@@ -255,6 +255,110 @@ resource "auth0_connection" "azure_ad" {
...
+				),
+			},
+			{
+				Config: random.Template(testAccConnectionOIDCConfigUpdate, rand),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.client_id", "1234567"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.client_secret", "1234567"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.#", "1"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.domain_aliases.3506632655", "example.com"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scope", "openid profile"),
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:298]:
```suggestion
```

File: auth0/resource_auth0_connection_test.go [hunk 9]
```diff
@@ -255,6 +255,110 @@ resource "auth0_connection" "azure_ad" {
...
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.issuer", "https://www.paypalobjects.com"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.jwks_uri", "https://api.paypal.com/v1/oauth2/certs"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.discovery_url", "https://www.paypalobjects.com/.well-known/openid-configuration"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.token_endpoint", "https://api.paypal.com/v1/oauth2/token"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.userinfo_endpoint", "https://api.paypal.com/v1/oauth2/token/userinfo"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.authorization_endpoint", "https://www.paypal.com/signin/authorize"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.#", "2"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.881205744", "email"),
+					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.4080487570", "profile"),
```
Comment (alexkappa) [inline auth0/resource_auth0_connection_test.go:308]:
```suggestion
					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.2517049750", "openid"),
					resource.TestCheckResourceAttr("auth0_connection.oidc", "options.0.scopes.881205744", "email"),
```

**Comment by incorrupt** [comment]
@alexkappa thanks for the suggestions, they came in handy as I wasn't able to run the `debug.DumpAttr()` command.
Everything is now passing the tests.

File: auth0/structure_auth0_connection.go [hunk 10]
```diff
@@ -116,6 +116,25 @@ func flattenConnectionOptionsSMS(o *management.ConnectionOptionsSMS) interface{}
 	}
 }
 
+func flattenConnectionOptionsOIDC(d Data, o *management.ConnectionOptionsOIDC) interface{} {
```
Comment (alexkappa) [inline auth0/structure_auth0_connection.go:119]: I think we don’t need the Data argument here right?

Comment (incorrupt) [reply auth0/structure_auth0_connection.go]: You are right!

**Pull request merged**